### PR TITLE
restbed: add restbed/4.8 recipe

### PR DESCRIPTION
**Pull request by madebr**
Specify library name and version:  **restbed/4.8**

Tested with MSVC2019, gcc@Linux, mingw@Linux

---

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/restbed/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                del self.options.fPIC
+            except Exception:
+                pass
+        if self.settings.os in ("Windows", ):
+            del self.options.ipc
+
+    def validate(self):
+        if self.settings.os in ("Windows", ) and not self.options.shared:
+            # The headers don't allow to avoid __declspec(dllimport)/__declsped(dllexport)
+            raise ConanInvalidConfiguration("Static libraries for Windows are not supported")
```
Comment (SpaceIm) [inline recipes/restbed/all/conanfile.py:52]: Indeed, it comes from https://github.com/Corvusoft/restbed/pull/434

File: recipes/restbed/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        copy(self, "LICENSE*", src=self.source_folder, dst=os.path.join(self.package_folder, "licenses"))
+        cmake = CMake(self)
+        cmake.install()
+        rm(self, "*.pdb", os.path.join(self.package_folder, "bin"))
+
+    @property
+    def _libzmq_target(self):
+        return "libzmq" if self.options.shared else "libzmq-static"
+
+    def package_info(self):
```
Comment (madebr) [inline recipes/restbed/all/conanfile.py:94]:
```suggestion
        rm(self, "*.pdb", os.path.join(self.package_folder, "bin"))

    def package_info(self):
```

File: recipes/restbed/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        if self.options.shared:
+            try:
+                del self.options.fPIC
+            except Exception:
+                pass
+        if self.settings.os in ("Windows", ):
+            del self.options.ipc
+
+    def validate(self):
+        if self.settings.os in ("Windows", ) and not self.options.shared:
```
Comment (uilianries) [inline recipes/restbed/all/conanfile.py:50]:
```suggestion
        if self.info.settings.os in ("Windows", ) and not self.info.options.shared:
```

File: recipes/restbed/all/conanfile.py (hunk 1 above)
Comment (madebr) [reply recipes/restbed/all/conanfile.py]: I fixed it with a regex.

File: recipes/restbed/all/conanfile.py (hunk 3 above)
Comment (madebr) [reply recipes/restbed/all/conanfile.py]: Fixed

File: recipes/restbed/all/conandata.yml [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+    url: "https://github.com/Corvusoft/restbed/archive/refs/tags/4.8.tar.gz"
+    sha256: "4801833f86a67b8a123c2c01203e259eb81157e1e9ef144a3b6395cb2d838a42"
+patches:
+  "4.8":
+    - patch_file: "patches/4.8-0001-cmake-conan.patch"
+      patch_description: "Use CMake targets + separate static/shared"
+      patch_type: "conan"
+    - patch_file: "patches/4.8-0002-mingw-deprecated-fix.patch"
+      patch_description: "MinGW apparently does not support deprecated and __declspec in the same line"
+      patch_type: "upstreamable"
```
Comment (uilianries) [inline recipes/restbed/all/conandata.yml:12]: What? It's not listed, please, update accordingly: https://github.com/conan-io/conan-center-index/blob/master/docs/conandata_yml_format.md#patch_type

Comment (madebr) [reply recipes/restbed/all/conandata.yml:12]:
```suggestion
      patch_type: "portability"
```

File: recipes/restbed/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
+from conan import ConanFile, Version
```
Comment (uilianries) [inline recipes/restbed/all/conanfile.py:1]:
```suggestion
from conan import ConanFile
from conan.tools.scm import Version
```

**Review by uilianries — APPROVED** [review]

File: recipes/restbed/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+import re
+
+required_conan_version = ">=1.52.0"
+
+
+class RestbedConan(ConanFile):
+    name = "restbed"
+    homepage = "https://github.com/Corvusoft/restbed"
+    description = "Corvusoft's Restbed framework brings asynchronous RESTful functionality to C++14 applications."
+    topics = ("restbed", "restful", "server", "client", "json", "http", "ssl", "tls")
```
Comment (jwillikers) [inline recipes/restbed/all/conanfile.py:17]: The package name need not be duplicated in topics.

```suggestion
    topics = ("restful", "server", "client", "json", "http", "ssl", "tls")
```

File: recipes/restbed/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    options = {
+        "shared": [True, False],
+        "fPIC": [True, False],
+        "ssl": [True, False],
+        "ipc": [True, False],
+    }
+    default_options = {
+        "shared": False,
+        "fPIC": True,
+        "ssl": True,
```
Comment (jwillikers) [inline recipes/restbed/all/conanfile.py:31]: `with_openssl` seems to be a more common option among Conan packages.

```suggestion
        "with_openssl": True,
```

File: recipes/restbed/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        if self.info.settings.compiler == "gcc":
+            if self.info.settings.compiler.version < Version(5):
+                raise ConanInvalidConfiguration("gcc 5+ is required for c++14 support")
+
+    def layout(self):
+        cmake_layout(self, src_folder="src")
+
+    def requirements(self):
+        self.requires("asio/1.24.0")
+        if self.options.ssl:
```
Comment (jwillikers) [inline recipes/restbed/all/conanfile.py:63]:
```suggestion
        if self.options.with_openssl:
```

File: recipes/restbed/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+            self.requires("openssl/3.0.5")
+
+    def source(self):
+        get(self, **self.conan_data["sources"][self.version],
+            destination=self.source_folder, strip_root=True)
+
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.variables["BUILD_TESTS"] = False
+        tc.variables["BUILD_SSL"] = self.options.ssl
```
Comment (jwillikers) [inline recipes/restbed/all/conanfile.py:73]:
```suggestion
        tc.variables["BUILD_SSL"] = self.options.with_openssl
```

File: recipes/restbed/all/test_package/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,11 @@
+cmake_minimum_required(VERSION 3.1)
+project(test_package LANGUAGES CXX)
+
+if(NOT CMAKE_CXX_STANDARD)
+    set(CMAKE_CXX_STANDARD 14)
+endif()
```
Comment (jwillikers) [inline recipes/restbed/all/test_package/CMakeLists.txt:6]: `CMAKE_CXX_STANDARD` should be set as a cache variable to allow it to be overridden. However, the use of target properties is preferred for setting the C++ standard when there is a minimum.

File: recipes/restbed/all/test_package/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+
+find_package(restbed REQUIRED CONFIG)
+
+add_executable(${PROJECT_NAME} test_package.cpp)
+target_link_libraries(${PROJECT_NAME} PRIVATE restbed::restbed)
```
Comment (jwillikers) [inline recipes/restbed/all/test_package/CMakeLists.txt:11]:
```suggestion
target_link_libraries(${PROJECT_NAME} PRIVATE restbed::restbed)
target_compile_features(${PROJECT_NAME} PRIVATE cxx_std_14)
```

File: recipes/restbed/all/test_package/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,11 @@
+cmake_minimum_required(VERSION 3.1)
```
Comment (jwillikers) [inline recipes/restbed/all/test_package/CMakeLists.txt:1]:
```suggestion
cmake_minimum_required(VERSION 3.8)
```

File: recipes/restbed/all/test_package/test_package.cpp [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    resource->set_path( "/resource" );
+    resource->set_method_handler( "POST", post_method_handler );
+
+    auto settings = std::make_shared< restbed::Settings >( );
+    settings->set_port( 1984 );
+    settings->set_default_header( "Connection", "close" );
+
+    restbed::Service service;
+    service.publish( resource );
+#if 0
```
Comment (jwillikers) [inline recipes/restbed/all/test_package/test_package.cpp:32]: Huh?

File: recipes/restbed/all/test_v1_package/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+conan_basic_setup(TARGETS)
+
+if(NOT CMAKE_CXX_STANDARD)
+    set(CMAKE_CXX_STANDARD 14)
+endif()
+
+find_package(restbed REQUIRED CONFIG)
+
+add_executable(${PROJECT_NAME} ../test_package/test_package.cpp)
+target_link_libraries(${PROJECT_NAME} PRIVATE restbed::restbed)
```
Comment (jwillikers) [inline recipes/restbed/all/test_v1_package/CMakeLists.txt:14]:
```suggestion
target_link_libraries(${PROJECT_NAME} PRIVATE restbed::restbed)
target_compile_features(${PROJECT_NAME} PRIVATE cxx_std_14)
```

File: recipes/restbed/all/test_v1_package/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,14 @@
+cmake_minimum_required(VERSION 3.1)
```
Comment (jwillikers) [inline recipes/restbed/all/test_v1_package/CMakeLists.txt:1]: cmake_minimum_required(VERSION 3.8)

**Review by jwillikers — CHANGES_REQUESTED** [review]

File: recipes/restbed/all/test_package/CMakeLists.txt (hunk 10 above)
Comment (madebr) [reply recipes/restbed/all/test_package/CMakeLists.txt]: Fixed in 12afe0dc8

File: recipes/restbed/all/test_package/CMakeLists.txt (hunk 12 above)
Comment (madebr) [reply recipes/restbed/all/test_package/CMakeLists.txt]: Fixed in 12afe0dc8

File: recipes/restbed/all/test_v1_package/CMakeLists.txt (hunk 14 above)
Comment (madebr) [reply recipes/restbed/all/test_v1_package/CMakeLists.txt:10]: Fixed in 12afe0dc8

File: recipes/restbed/all/test_package/test_package.cpp (hunk 13 above)
Comment (madebr) [reply recipes/restbed/all/test_package/test_package.cpp]: Fixed in 12afe0dc8

File: recipes/restbed/all/conanfile.py (hunk 9 above)
Comment (madebr) [reply recipes/restbed/all/conanfile.py]: Fixed in 12afe0dc8

File: recipes/restbed/all/conanfile.py (hunk 7 above)
Comment (madebr) [reply recipes/restbed/all/conanfile.py]: Fixed in 12afe0dc8

**Review by jwillikers — DISMISSED** [review]

File: recipes/restbed/all/conanfile.py [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+            try:
+                del self.options.fPIC
+            except Exception:
+                pass
+        if self.settings.os in ("Windows", ):
+            del self.options.ipc
+
+    def validate(self):
+        if getattr(self.info.settings.compiler, "cppstd"):
+            check_min_cppstd(self, 14)
```
Comment (toge) [inline recipes/restbed/all/conanfile.py:53]: Please add other compiler(clang, msvc, apple-clang) version checks like template:
https://github.com/conan-io/conan-center-index/blob/master/docs/package_templates/cmake_package/all/conanfile.py#L90-L96

Comment (madebr) [reply recipes/restbed/all/conanfile.py]: Addressed in 8e47eb107

**Review by jwillikers — APPROVED** [review]

File: recipes/restbed/all/conanfile.py [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+required_conan_version = ">=1.52.0"
+
+
+class RestbedConan(ConanFile):
+    name = "restbed"
+    homepage = "https://github.com/Corvusoft/restbed"
+    description = "Corvusoft's Restbed framework brings asynchronous RESTful functionality to C++14 applications."
+    topics = ("restful", "server", "client", "json", "http", "ssl", "tls")
+    url = "https://github.com/conan-io/conan-center-index"
+    license = "AGPL-3.0-or-later", "CPL"  # Corvusoft Permissive License (CPL)
```
Comment (prince-chrismc) [inline recipes/restbed/all/conanfile.py:20]: https://spdx.org/licenses/CPL.html says 404 🤔 

we are doing https://github.com/conan-io/conan-center-index/blob/master/docs/faqs.md#what-license-should-i-use-for-a-custom-project-specific-license now 🙏

File: recipes/restbed/all/conandata.yml [hunk 18]
```diff
@@ -0,0 +1,12 @@
+sources:
+  "4.8":
+    url: "https://github.com/Corvusoft/restbed/archive/refs/tags/4.8.tar.gz"
+    sha256: "4801833f86a67b8a123c2c01203e259eb81157e1e9ef144a3b6395cb2d838a42"
+patches:
```
Comment (prince-chrismc) [inline recipes/restbed/all/conandata.yml:5]: This project is active, might be worth tossing up a PR 😉

**Review by prince-chrismc — COMMENTED** [review]
One nit about license, otherwise it's good

File: recipes/restbed/all/conanfile.py (hunk 17 above)
Comment (madebr) [reply recipes/restbed/all/conanfile.py:20]:
```suggestion
    license = "AGPL-3.0-or-later", "LicenseRef-CPL"  # Corvusoft Permissive License (CPL)
```

**Review by jwillikers — APPROVED** [review]

File: recipes/restbed/all/conandata.yml (hunk 18 above)
Comment (madebr) [reply recipes/restbed/all/conandata.yml:5]: https://github.com/Corvusoft/restbed/pull/532

**Review by prince-chrismc — APPROVED** [review]

**Pull request merged**